RM-9079 Add ASP.NET Core extension methods for static resource caching

a-ctor opened the pull request:
https://re-motion.atlassian.net/browse/RM-9079

MichaelKetting commented on Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs:
```diff
@@ -0,0 +1,124 @@
+    if (indexOfSlash == -1)
+      return false;
+
+    var restPath = requestPathEndsWithSlash
```
"rest" -> "remaining"

MichaelKetting commented on Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs:
```diff
@@ -0,0 +1,124 @@
+        ? requestPath
+        : requestPath + new PathString("/");
+
+    if (!requestPathWithTrailingSlash.StartsWithSegments(_requestPath, out var rest) || !rest.HasValue)
```
rest -> remaining

MichaelKetting commented on Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs:
```diff
@@ -0,0 +1,124 @@
+  /// Tries to retrieve a cache key from <paramref name="context"/>, which was previously removed using <see cref="RemoveCacheKeyMiddleware"/>.
+  /// The cache key matches the value returned by <see cref="IStaticResourceCacheKeyProvider"/>.
+  /// </summary>
+  public static bool TryGetRemovedCacheKey (HttpContext context, [NotNullWhen(true)] out string? cacheKey)
```
do we ever need the actual value or would this be clearer as a simple boolean operation?

a-ctor replied:
We don't need it but I think it is a good to leave options open so I would keep it here.

a-ctor commented on Remotion/Web/Core/Resources/CachingApplicationBuilderExtensions.cs:
```diff
@@ -0,0 +1,69 @@
+  /// <remarks>
+  /// Requires <see cref="ResourcePathBuilderBasedStaticResourceCacheKeyProvider"/> to be useful.
+  /// </remarks>
+  public static IApplicationBuilder UseRemotionStaticFiles (this IApplicationBuilder builder, StaticFileOptions options, int cacheMaxAge = 604_800)
```
Looking at this again I chose to introduce own own settings object `StaticFileCachingOptions` with init-only properties. I also converted the max age to a `TimeSpan`, and move the `Cache-Control` header generation over there. This makes the options more explicit and clear, while allowing the customization of the header value via subclassing.

a-ctor commented on Remotion/Web/Core/Resources/RemoveCacheKeyMiddleware.cs:
```diff
@@ -0,0 +1,124 @@
+/// <remarks>
+/// The preferred way of using this middleware is <see cref="CachingApplicationBuilderExtensions"/>.<see cref="CachingApplicationBuilderExtensions.UseRemotionStaticFiles"/>.
+/// </remarks>
+public class RemoveCacheKeyMiddleware
```
It is, otherwise it would not compile :D

a-ctor commented on Remotion/Web/Core/Resources/CachingApplicationBuilderExtensions.cs:
```diff
@@ -0,0 +1,69 @@
+  /// Only removes the cache key from the URL but does not change the request in any other way.
+  /// Other middleware logic can check if a cache key was removed using <see cref="RemoveCacheKeyMiddleware"/>.<see cref="RemoveCacheKeyMiddleware.TryGetRemovedCacheKey"/>.
+  /// </remarks>
+  public static IApplicationBuilder UseRemoveCacheKeyMiddleware (this IApplicationBuilder builder, PathString requestPath)
```
Yeah probably does not have a good reason to exist anymore as most of the customization is possible in the wrapped `OnPrepareResponse` anyways.

a-ctor commented:
@MichaelKetting I have addressed all of your review comments. I also discovered that the `UseRemotionStaticFilesCacheKeyRemovalMiddleware` was better of being a rewrite rule than a middleware. In terms of code the changes are minimal but it makes it easier to reuse if needed in the future. Another benefit is that we now use the official rewrite middleware, which handles edge cases like routing for us.

a-ctor commented on Remotion/Web/Core/Resources/CachingApplicationBuilderExtensions.cs:
```diff
@@ -0,0 +1,57 @@
+    ArgumentUtility.CheckNotNull(nameof(builder), builder);
+    ArgumentUtility.CheckNotNull(nameof(options), options);
+
+    builder.UseRewriter(new RewriteOptions().Add(new RemotionStaticFilesCacheKeyRemovalRewriteRule(options.RequestPath, true)));
```
It is possible to use the rewrite middleware multiple times. Web Tests are currently not possible because we don't run the test site with .NET, which is required for the middleware stuff to run.

MichaelKetting replied:
> Web Tests are currently not possible because we don't run the test site with .NET, which is required for the middleware stuff to run.

Hence please create a ticket for re-motion 7 :)

MichaelKetting replied:
https://re-motion.atlassian.net/browse/RM-9132

a-ctor commented on Remotion/Web/UnitTests/Core/Resources/RemotionStaticFilesCacheKeyRemovalRewriteRuleTest.cs:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  [Test]
+  [TestCase("/static/cache_/", "/static/", "")]
+  [TestCase("/static/cache_abc/", "/static/", "abc")]
```
The rewrite rule tested in this file does not do anything cache related on its own. The static file middleware will only handle the request (and add caching) if the file exists. Otherwise, further middlewares in the chain are called. As such, `/static/` would not be handled by the static file middleware as it is not a file and will fail with 404 without adding any caching header.

MichaelKetting commented on Remotion/Web/UnitTests/Core/Resources/RemotionStaticFilesCacheKeyRemovalRewriteRuleTest.cs:
```diff
@@ -0,0 +1,94 @@
+public class RemotionStaticFilesCacheKeyRemovalRewriteRuleTest
+{
+  [Test]
+  public void Constructor_RequestPathWithNoTrailingSlash_Throws ()
+  {
+    Assert.That(
+        () => new RemotionStaticFilesCacheKeyRemovalRewriteRule("/abc/", false),
+        Throws.ArgumentException.With.ArgumentExceptionMessageEqualTo(
+            "Request path must not end in a slash.",
+            "requestPath"));
+  }
```
You want to also add a test for the empty case?
